File: src/main/java/dev/haarolean/gnutardbot/abilities/CheckLinksAbility.java
package dev.haarolean.gnutardbot.abilities;

import dev.haarolean.gnutardbot.TardBot;
import dev.haarolean.gnutardbot.util.MemberUtils;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;
import org.telegram.abilitybots.api.objects.*;
import org.telegram.telegrambots.meta.api.methods.groupadministration.BanChatMember;
import org.telegram.telegrambots.meta.api.methods.updatingmessages.DeleteMessage;

import static org.springframework.beans.factory.config.BeanDefinition.SCOPE_PROTOTYPE;
import static org.telegram.abilitybots.api.objects.Flag.EDITED_MESSAGE;
import static org.telegram.abilitybots.api.util.AbilityUtils.isUserMessage;

@Component
@Scope(SCOPE_PROTOTYPE)
public class CheckLinksAbility implements AbilityProvider {

    private final MemberUtils memberUtils;
    private final TardBot bot;

    public CheckLinksAbility(TardBot bot) {
        this.bot = bot;
        this.memberUtils = new MemberUtils(bot);
    }

    @Override
    public Ability buildAbility() {
        return Ability
                .builder()
                .name("default")
                .locality(Locality.ALL)
                .privacy(Privacy.PUBLIC)
                .action(ctx -> {
                    if (!shouldProceed(ctx)) return;
                    checkLinks(ctx);
                })
                .build();
    }

    private boolean shouldProceed(MessageContext ctx) {
        // a shitty fix for a more shitty "bug":
        // https://github.com/rubenlagus/TelegramBots/issues/1296
        var update = ctx.update();
        if (!isUserMessage(update)) return true;
        if (Flag.MESSAGE.test(update) || EDITED_MESSAGE.test(update)) return false;
        return false;
    }

    private void checkLinks(MessageContext ctx) {
        var message = ctx.update().getMessage();
        var sender = message.getFrom();
        long senderId = sender.getId();
        var chatId = message.getChatId().toString();

        if (bot.isAdmin(senderId)) return;
        if (bot.isGroupAdmin(ctx.update(), senderId)) return;
        if (memberUtils.isUserKnown(chatId, senderId)) return;

        var hasLink = message.getText().matches(".*https?://.*");

        if (!hasLink) return;

        var messageText = "Yo nibba, you triggered an anti spam thingy. " +
                "No links are allowed if you're not in the chat group. " +
                "Contact the admins if you're not a bot.";
        var senderTag = !sender.getUserName().isEmpty() ? ("@" + sender.getUserName() + " ") : "";
        var replyMessage = bot.silent().forceReply(senderTag + messageText, message.getChatId());

        var banRequest = BanChatMember
                .builder()
                .chatId(chatId)
                .userId(senderId)
                .build();

        bot.silent().execute(banRequest);
        replyMessage.ifPresent(msg ->
                bot.silent().execute(new DeleteMessage(chatId, msg.getMessageId())));
    }
}

File: src/main/kotlin/dev/haarolean/gnutardbot/abilities/CheckLinksAbility.kt
package dev.haarolean.gnutardbot.abilities

import dev.haarolean.gnutardbot.TardBot
import dev.haarolean.gnutardbot.util.MemberUtils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import org.springframework.beans.factory.config.BeanDefinition
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Component
import org.telegram.abilitybots.api.objects.MessageContext
import org.telegram.abilitybots.api.util.AbilityUtils
import org.telegram.telegrambots.meta.api.methods.send.SendMessage
import org.telegram.telegrambots.meta.api.methods.updatingmessages.DeleteMessage
import org.telegram.telegrambots.meta.api.objects.Message
import kotlin.time.Duration.Companion.seconds

@Component
@Scope(BeanDefinition.SCOPE_PROTOTYPE)
class CheckLinksAbility(private val bot: TardBot) : AbilityHandler {
    private val memberUtils: MemberUtils = (MemberUtils(bot))

    override fun isApplicable(ctx: MessageContext): Boolean {
        val update = ctx.update()
        if (AbilityUtils.isUserMessage(update)) return false
        val message: Message = when {
            update.hasMessage() -> update.message
            update.hasEditedMessage() -> update.editedMessage
            else -> return false
        }
        val sender = message.from
        val senderId = sender.id
        val chatId = message.chatId.toString()
        if (bot.isAdmin(senderId)) return false
        if (bot.isGroupAdmin(ctx.update(), senderId)) return false
        if (memberUtils.isUserKnown(chatId, senderId)) return false

        if (!message.hasEntities()) return false

        return message.entities.stream().anyMatch { it.type == "url" || it.type == "text_link" }
    }

    override fun handle(ctx: MessageContext) {
        checkLinks(ctx)
    }

    private fun checkLinks(ctx: MessageContext) {
        var message = ctx.update().message
        if (!ctx.update().hasMessage()) message = ctx.update().editedMessage
        val chatId = message.chatId.toString()

        val sender = message.from
        val senderName = escapeMarkdownV2(sender.firstName + " " + (sender.lastName ?: ""))
        val mention = "[${senderName}](tg://user?id=${message.from.id})"

        val messageText = "$mention, the link has been removed" +
                " as no links are allowed if you're not a member of the chat group"

        val replyRequest = SendMessage
            .builder()
            .replyToMessageId(message.messageId)
            .chatId(chatId)
            .text(messageText)
            .parseMode("MarkdownV2")
            .build()
        val replyMessage = bot.silent().execute(replyRequest)

        bot.silent().execute(DeleteMessage(chatId, message.messageId))

        replyMessage.ifPresent {
            CoroutineScope(Dispatchers.IO).launch {
                delay(60.seconds)
                bot.silent().execute(DeleteMessage(chatId, it.messageId))
            }
        }
    }

    fun escapeMarkdownV2(text: String): String {
        val specialCharacters = "_*[]()~`>#+-=|{}.!"
        return text.map { if (it in specialCharacters) "\\$it" else it }.joinToString("")
    }

}
